### ISSUE TITLE
[Feature] Allow pl.aiv_shard on a vector-produced Vec tile so a per-lane scattered gather_row can be expressed inside pl.split_aiv

### ISSUE THREAD

**Issue by Hzfengsy** [hw-native-sys/pypto#2244]
## Background

In pypto-lib, `models/deepseek/v4-flash/decode_sparse_attn.py` (`qk_pv`) on `a2a3` has a **compile** problem: there is no way to express, at tensor level, "the two AIV lanes each gather half of a scattered row set into UB, then hand the assembled tile to the cube as a matmul B-operand".

`qk_pv` assembles each `[ATTN_K_TILE, HEAD_DIM]` sparse KV tile with 128 single-row `pl.gather_row` calls on the **cube**. Measured on 910B1, the cube's per-row `GM->L1` ND2NZ sustains only ~3.5 scattered 1 KB loads in flight (~21 us for the tile) while a vector lane reaches ~13 (~5 us), and `qk_pv` is 73% MTE2-occupied while its two AIV lanes idle at ~7%. Moving the gather to the AIV lanes is worth ~4% end-to-end on the layer.

That works today only via a GM staging buffer plus hand-written `pl.system.set_ffts` + `sync_set`/`sync_wait` (the author-side barrier that #1911 describes). The tensor-level `split_aiv` + `aic_gather` formulation, which needs neither, is blocked.

Reproduce with:

    python repro_aiv_shard_gather_row.py --fix 0   # the original rejection
    python repro_aiv_shard_gather_row.py --fix 1   # "derive from aiv_shard"
    python repro_aiv_shard_gather_row.py --fix 3   # "let the implicit path halve it"

The repro case is **not committed** — save the file below at the pypto-lib repo root to run it. Files required: `repro_aiv_shard_gather_row.py` (self-contained; imports only `pypto.language` and pypto-lib's `golden` harness).

<details>
<summary>Reproduction case — <code>repro_aiv_shard_gather_row.py</code> (save at the pypto-lib repo root)</summary>

```python
"""Repro: no tensor-level way to split a scattered GM->UB gather across the two AIV lanes.

Goal: the two AIV lanes each gather half of a scattered row set into UB, then the
assembled [ROWS, HEAD_DIM] tile is handed to the cube as a matmul B-operand. The
scattered row indices are runtime values (a paged KV top-k list), so the gather
must accept arbitrary per-row source offsets.

`pl.paged_gather(space=Vec)` is the only tensor-level op that produces a UB-backed
Tensor, but it cannot be given a per-lane sub-range of the index list:
  * a sliced index tensor (static or dynamic) -> "indices must be TensorType"
  * a pl.aiv_shard(...) result                -> "indices must be TensorType"
  * a [2, n/2] per-lane layout               -> "requires 2D indices to have shape [1, n]"

`pl.gather_row` does take arbitrary runtime offsets, and `pl.full` gives it a
UB-backed Tensor accumulator. But putting that inside `pl.split_aiv` is rejected,
and each of the three fixes LowerAutoVectorSplit suggests is itself blocked.

Run each variant from the pypto-lib repo root:
    python repro_aiv_shard_gather_row.py --fix 0   # the original rejection
    python repro_aiv_shard_gather_row.py --fix 1   # "derive from aiv_shard"
    python repro_aiv_shard_gather_row.py --fix 3   # "let the implicit path halve it"
"""

import pypto.language as pl

POOL_ROWS = 16384
HEAD_DIM = 512
ROWS = 128
HALF = ROWS // 2
Q_ROWS = 16


@pl.jit
def fix0_half_extent(
    pool: pl.Tensor[[POOL_ROWS, HEAD_DIM], pl.BF16],
    idx: pl.Tensor[[ROWS], pl.INT32],
    q: pl.Tensor[[Q_ROWS, HEAD_DIM], pl.BF16],
    out: pl.Out[pl.Tensor[[Q_ROWS, ROWS], pl.FP32]],
):
    """Half-extent UB accumulator; the gather's SOURCE offset references aiv."""
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="fix0", allow_early_resolve=True):
        for aiv in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
            ub = pl.full([HALF, HEAD_DIM], dtype=pl.BF16, value=0.0)
            for k in pl.range(HALF):
                src = pl.cast(pl.read(idx, [aiv * HALF + k]), pl.INDEX)
                ub = pl.gather_row(ub, pool, [k, 0], [src, 0], [1, HEAD_DIM])
            kv = pl.aic_gather(ub)
        out[0:Q_ROWS, 0:ROWS] = pl.matmul(q, kv, b_trans=True, out_dtype=pl.FP32)
    return out


@pl.jit
def fix1_aiv_shard(
    pool: pl.Tensor[[POOL_ROWS, HEAD_DIM], pl.BF16],
    idx: pl.Tensor[[ROWS], pl.INT32],
    q: pl.Tensor[[Q_ROWS, HEAD_DIM], pl.BF16],
    out: pl.Out[pl.Tensor[[Q_ROWS, ROWS], pl.FP32]],
):
    """Fix (1): derive the accumulator from a pl.aiv_shard result."""
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="fix1", allow_early_resolve=True):
        for aiv in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
            ub_full = pl.full([ROWS, HEAD_DIM], dtype=pl.BF16, value=0.0)
            ub = pl.aiv_shard(ub_full)
            for k in pl.range(HALF):
                src = pl.cast(pl.read(idx, [aiv * HALF + k]), pl.INDEX)
                ub = pl.gather_row(ub, pool, [k, 0], [src, 0], [1, HEAD_DIM])
            kv = pl.aic_gather(ub)
        out[0:Q_ROWS, 0:ROWS] = pl.matmul(q, kv, b_trans=True, out_dtype=pl.FP32)
    return out


@pl.jit
def fix3_implicit(
    pool: pl.Tensor[[POOL_ROWS, HEAD_DIM], pl.BF16],
    idx: pl.Tensor[[ROWS], pl.INT32],
    q: pl.Tensor[[Q_ROWS, HEAD_DIM], pl.BF16],
    out: pl.Out[pl.Tensor[[Q_ROWS, ROWS], pl.FP32]],
):
    """Fix (3): no explicit boundary ops; both offsets reference the lane index."""
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="fix3", allow_early_resolve=True):
        for aiv in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
            ub = pl.full([ROWS, HEAD_DIM], dtype=pl.BF16, value=0.0)
            for k in pl.range(HALF):
                src = pl.cast(pl.read(idx, [aiv * HALF + k]), pl.INDEX)
                ub = pl.gather_row(ub, pool, [aiv * HALF + k, 0], [src, 0], [1, HEAD_DIM])
            kv = ub
        out[0:Q_ROWS, 0:ROWS] = pl.matmul(q, kv, b_trans=True, out_dtype=pl.FP32)
    return out


VARIANTS = {0: fix0_half_extent, 1: fix1_aiv_shard, 3: fix3_implicit}


def golden(tensors):
    tensors["out"][:] = (
        tensors["q"].float() @ tensors["pool"][tensors["idx"].long()].float().t()
    ).to(tensors["out"].dtype)


def build_tensor_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec("pool", [POOL_ROWS, HEAD_DIM], torch.bfloat16,
                   init_value=lambda: torch.randn(POOL_ROWS, HEAD_DIM)),
        TensorSpec("idx", [ROWS], torch.int32,
                   init_value=lambda: ((torch.arange(ROWS, dtype=torch.int64) * 977)
                                       % (POOL_ROWS - ROWS)).to(torch.int32)),
        TensorSpec("q", [Q_ROWS, HEAD_DIM], torch.bfloat16,
                   init_value=lambda: torch.randn(Q_ROWS, HEAD_DIM) / HEAD_DIM ** 0.5),
        TensorSpec("out", [Q_ROWS, ROWS], torch.float32, is_output=True),
    ]


if __name__ == "__main__":
    import argparse
    from golden import run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument("--fix", type=int, default=0, choices=sorted(VARIANTS))
    parser.add_argument("-p", "--platform", type=str, default="a2a3")
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()

    result = run_jit(
        fn=VARIANTS[args.fix],
        specs=build_tensor_specs(),
        golden_fn=golden,
        runtime_cfg=dict(platform=args.platform, device_id=args.device),
        rtol=1e-2, atol=1e-2,
    )
    print("PASS" if result.passed else f"FAIL {result.error}")
```

</details>

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `14f3863` (branch: `perf/assemble-swa-decode-window-in-l1-drop-gm-gather`) |
| pypto | `6234d0e9` (branch: `main`) |
| simpler | `9922afdb` (detached) |
| ptoas | `0.54` |
| pto-isa | `83d01313` |
| CANN | `8.5.1` (runs), `8.3.RC1.alpha003` at `ascend-toolkit/latest` |

Version mismatch (off pinned): **simpler** expected `dccb8379080b43173744b9981de2542b3d025e19`, actual `9922afdb08cc6f203eaf39328661e2f2648d333d`. This is a **compile-time** failure in `LowerAutoVectorSplit` / `ConvertTensorToTileOps`; simpler is the runtime and takes no part in those passes, so the divergence is not believed to affect the result. ptoas and pto-isa are on their pins.

Diagnosis: **pypto** — all three failures are raised by pypto compiler passes and verifiers.

### Summary

`pl.gather_row` is the only op that gathers scattered GM rows at arbitrary **runtime** offsets, and `pl.full` is the only tensor-level way to give it a UB-backed `Tensor` accumulator. That pair cannot be placed inside a `pl.split_aiv` region: `LowerAutoVectorSplit` classifies `tile.gather_row` as full-width and names three fixes, and **each of the three is independently blocked**.

| variant | what it does | result |
|---|---|---|
| `--fix 0` | half-extent `pl.full`, gather's **source** offset references `aiv_id`, `aic_gather` | `LowerAutoVectorSplit: a pl.split_aiv region mixes explicit tile.aiv_shard/tile.aic_gather boundary ops with plain full-width vector op(s) [tile.gather_row] ...` |
| `--fix 1` | full-extent `pl.full` -> `pl.aiv_shard` -> gather -> `aic_gather` | `AivSplitValid`: *"'tile.aiv_shard' operand is in Vec, but it transfers a cube-produced value across the cross-core boundary and requires Acc. ... A Vec operand is vector-produced (pl.load / pl.full) and already lives on the AIV lane — drop the pl.aiv_shard and let the implicit affinity-gated split halve it."* |
| `--fix 3` | full-extent `pl.full`, **both** offsets reference `aiv_id`, no boundary ops | `AllocatedMemoryAddr`: `Function 'fix3_aic': Mat buffer usage (1064960 bytes) exceeds platform limit (524288 bytes)` |

Two observations:

1. In `--fix 0` the gather's **source** offset already references `aiv_id`, which is what fix (2) asks for ("an op whose arguments reference `aiv_id` is per-lane by construction and is accepted"). It is still rejected — the check appears to consider only the destination/accumulator, not the source.
2. Fix (1) is not applicable to this shape at all: `aiv_shard` is specified for a **cube-produced** (`Acc`) value crossing C2V, but the accumulator here is vector-produced (`Vec`). The diagnostic itself redirects to fix (3), which then overflows L1.

The net effect is that a per-lane scattered gather has no tensor-level spelling, and the only working formulation is a GM staging buffer with hand-written `set_ffts` + `sync_set`/`sync_wait`.

### Motivation / Use Case

DeepSeek-V4-Flash CSA decode. `qk_pv` is 29.7% of the layer's machine core-time and 73% MTE2-occupied on the cube, gathering 128 scattered 1 KB KV rows per tile; its AIV lanes are ~7% occupied. Micro-benchmarks on 910B1 (same binary, disjoint index sets, ordering-controlled):

| path | us for one 128x512 BF16 tile |
|---|---|
| cube, 128 x `gather_row` (today) | 20.90 |
| AIV pair, UB accumulate + coalesced store | 3.26 |

The GM-staging workaround lands ~4% end-to-end on the layer, but it costs a 4 MB scratch buffer, triples the bytes moved (scattered read + write + read), and needs the manual barrier. `aic_gather` returning a `Tensor` in `Mat` — verified: with a `Tensor` input it returns `pl.Tensor[[256, 512], pl.BF16]` — would remove the staging buffer, the extra traffic, and the barrier in one step.

### Proposed API / Behavior

Preferred (this is fix (1) made applicable): let **`pl.aiv_shard` accept a vector-produced `Vec` operand**, sharding a locally-built UB tile into per-lane halves. Today it is restricted to `Acc` (cube-produced, C2V). Accepting `Vec` would let a `pl.full` accumulator be sharded and then filled by per-lane `gather_row` calls, with `aic_gather` carrying the assembled tile to the cube:

```python
for aiv in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
    ub = pl.aiv_shard(pl.full([ROWS, HEAD_DIM], dtype=pl.BF16, value=0.0))  # -> [ROWS/2, HEAD_DIM]
    for k in pl.range(ROWS // 2):
        src = pl.cast(pl.read(idx, [aiv * (ROWS // 2) + k]), pl.INDEX)
        ub = pl.gather_row(ub, pool, [k, 0], [src, 0], [1, HEAD_DIM])
    kv = pl.aic_gather(ub)                                                  # -> [ROWS, HEAD_DIM] in Mat
out = pl.matmul(q, kv, b_trans=True, out_dtype=pl.FP32)
```

Alternatives that would also unblock it:

- **Honour fix (2) for DMA ops** — accept `tile.gather_row` as per-lane when *any* of its offsets (source included) references the region's lane index, not only the destination.
- **Let `paged_gather` take a per-lane index range** — a sliced/sharded index tensor, or an explicit index-offset argument. Today `pl.paged_gather` is the only tensor-level op producing a UB `Tensor`, but it rejects a sliced index tensor (static *or* dynamic) with `indices must be TensorType`, rejects a `pl.aiv_shard` result with the same error, and rejects a `[2, n/2]` per-lane layout with `requires 2D indices to have shape [1, n], but got first dim 2`. Any one of these would make the `paged_gather` + `aic_gather` form work directly.

### Alternatives Considered

- **GM staging + explicit barrier (what pypto-lib ships today).** The AIV lanes coalesce chunks into a GM scratch and the cube reloads them contiguously, ordered by `pl.system.set_ffts` + `sync_set(pipe=MTE3, core_type="aiv")` / `sync_wait(pipe=MTE2, core_type="aic")`. Correct and ~4% faster end-to-end, but needs a 4 MB scratch, triples bytes moved, and relies on the hand-written barrier that #1911 wants automated. Without the barrier the kernel passes 9/9 uninstrumented runs and fails 3/3 under `--enable-l2-swimlane 1` — a silent race, exactly the failure mode #1911 describes.
- **Rewrite the qk_pv consumer chain at tile level.** `pl.tile.gather_row` accepts a `Vec` destination and arbitrary runtime offsets, so a `create_tile` + `tile.gather_row` + `aic_gather` chain works — but `aic_gather` then returns a `Tile`, and `pl.matmul` rejects mixed `Tensor`/`Tile` operands, so the whole QK/softmax/PV chain (~25 lines) has to move to tile level. Possible today; avoided because it makes the kernel materially harder to read for what should be an operand-plumbing detail.

### Additional Context

Related:
- #1911 — auto-fence AIV->AIC (V2C) GM read-after-write. The barrier this issue is trying to avoid needing is exactly the one #1911 proposes to automate; the shipped pypto-lib workaround is the "status quo" alternative listed there.
- #2203 — `LowerAutoVectorSplit` halves a vector op's result without sharding its full-width operand. Same pass, adjacent defect family (that one is about elementwise ops with un-sharded operands; this one is about a DMA op being classified full-width).
- #1820 — RFC for the explicit `pl.split_aiv` + `aiv_shard`/`aic_gather` DSL this issue exercises.
- #2130 — `pl.pipeline` cannot generate skewed cross-core schedules without manual `sync_set`/`sync_wait`. Consistent with what was measured here: `pl.pipeline(stage=2)` over the staging chunks gave no overlap (349.8 vs 348.8 us) because the manual per-chunk handover already serialises the pair.
- #2006 — `pl.split_aiv(2, SplitMode.NONE)` silently drops half the work.

Measurements above are from 910B1, CANN 8.5.1, `--start-pos 8192`, `B=4 S=2 T=8`.
